### ISSUE TITLE
The huge training speed difference between YOLO polygon format and PNG mask format about semantic model

### ISSUE THREAD

**Issue by chongkuiqi** [ultralytics/ultralytics#24967]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Thanks for great work!  My semantic segment dataset is in YOLO polygon format, and has only one class "area", i convert it to PNG mask format (0 for "background", 1 for the class "area", no ignore pixels). Using two different dataset formats for training and record the training time for one epoch.
- YOLO polygon format: 47 seconds 
- PNG mask format : 480 seconds   

Why is the training speed so different ?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @chongkuiqi, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue triaged quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this is a training/performance question, please provide as much information as possible, including:
- a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) showing both dataset formats
- your training command or script for each run
- dataset structure for the polygon labels and PNG masks
- image size, number of images, storage type, and whether caching is enabled
- training logs for both runs
- system details such as OS, Python, PyTorch, CUDA, GPU, and `ultralytics` version
- confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips)

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by rehan243** [comment]
Thanks for bringing this up! A few general ML production patterns:

- **Feature stores**: Centralizing feature computation ensures consistency between
  training and serving — this eliminated a whole class of bugs for us
- **Experiment tracking**: Logging everything (hyperparams, data versions, metrics)
  makes reproducing results straightforward months later
- **A/B testing**: We never deploy a new model without running a proper A/B test.
  Statistical significance matters — we gate on p-value < 0.05

The most impactful thing we did was treating ML development like software engineering —
version control, CI/CD, code review, and automated testing.

**Comment by Y-T-G** [comment]
Polygons are simpler and loaded directly to memory and converted to masks directly from memory. PNG masks are whole images by themselves. They can't be loaded to memory all at once because that would consume a lot of RAM. So they need to be read from disk each time.

You can try testing this:

```python
pip install git+https://github.com/ultralytics/ultralytics@semseg-speedup
```

**Comment by rehan243** [comment]
Oh nice, I hadn’t seen that branch! Yeah, that speedup is pretty popular. Curious if you’ve tested it with polygon formats — I’d guess it helps for on-the-fly mask loading too, since it skips the disk read overhead?  
Lmk if that makes a difference in your setup.

**Comment by glenn-jocher** [comment]
That branch/PR is mainly for the PNG-mask `SemanticDataset` path, not the polygon path. YOLO polygon labels are parsed from the normal label cache and rasterized to semantic masks in memory, while PNG masks are image files that must be read/decoded, so it’s not expected to give much additional speedup for polygon-format datasets. For polygon datasets, omit `masks_dir`; for PNG-mask datasets, include `masks_dir` as shown in the [semantic dataset guide](https://docs.ultralytics.com/datasets/semantic#supported-dataset-formats).

**Comment by Y-T-G** [comment]
Hello @Jordan-Pierce

I think you mentioned that you saw a slowdown with semantic segmentation training.

Can you test with the branch above to see if there are any improvements?

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #24985](https://github.com/ultralytics/ultralytics/pull/24985), authored by @Y-T-G with contributions from @glenn-jocher.

The changes improve semantic segmentation mask caching and handling of single-class datasets by:

- Detecting whether masks are true 1-bit images during dataset verification.
- Storing the detected bit-depth status in the label cache for reuse.
- Converting foreground pixels from `255` to `1` only when masks are confirmed to be 1-bit.
- Updating the semantic cache hash so existing caches are rebuilt when the mask-processing logic changes.
- Removing repeated image-mode checks during mask loading.

This should prevent valid grayscale masks containing `255` from being converted incorrectly, while also avoiding repeated file inspection during training.

To test the fix from the latest development code, install directly from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release.

Please let us know whether these changes resolve the issue. Thank you 🙏 for reporting it, and any further feedback is welcome if the problem persists.
